### PR TITLE
Feature/hunger

### PR DESCRIPTION
**Pull request by plally**
adds a simple hud that displays health, hunger, and thirst as well as a system for gaining/losing thirst and hunger

### CODE REVIEW

File: gamemode/modules/hunger/sh_hunger.lua [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+function PlayerMeta:GetThirst()
+    return self:GetNWInt("rvr_thirst")
+end
+
+function PlayerMeta:SetThirst()
+    self:SetNWInt("rvr_thirst")
+end
+
+function PlayerMeta:AddThirst(n)
```
Comment (plally) [inline gamemode/modules/hunger/sh_hunger.lua]: I dont like calling this thirst and hunger. this value isnt their Thirst/Hunger its their Water/Food. Calling it "Food" and "Water" seems weird though, any ideas

Comment (samuelWilliams99) [reply gamemode/modules/hunger/sh_hunger.lua]: I think `AddFood` and `AddWater` is fine?

**Comment by samuelWilliams99** [comment]
Will the gui for this be a different branch then?

**Comment by plally** [comment]
i didnt bother doing a hud on this branch as the hud will have to have more than thirst/hunger

**Comment by samuelWilliams99** [comment]
Fair

**Comment by samuelWilliams99** [comment]
No AddCSLuaFile for ur shared file?

File: gamemode/modules/hunger/sh_hunger.lua [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    return self:GetNWInt("rvr_food")
+end
+
+function PlayerMeta:SetFood(n)
+    self:SetNWInt("rvr_food", n)
+end
+
+function PlayerMeta:AddFood(n)
+    local food = self:GetNWInt("rvr_food")
+    food = math.max( food+n, 0 )
```
Comment (samuelWilliams99) [inline gamemode/modules/hunger/sh_hunger.lua]: bruh spaces

File: gamemode/modules/hunger/sv_hunger.lua [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    ply:SetWater( MAX_WATER )
+end )
+
+timer.Create( "rvr_hunger_subtract", DELAY, 0, function()
+    for _, ply in pairs( player.GetHumans() ) do
+        
+        ply:AddFood( FOOD_LOSS_AMOUNT )
+        ply:AddWater( WATER_LOSS_AMOUNT )
+
+        -- TODO seperate timer
```
Comment (samuelWilliams99) [inline gamemode/modules/hunger/sv_hunger.lua]: Yeah the damage timer should be different

File: gamemode/modules/hunger/sh_hunger.lua [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+function PlayerMeta:GetFood()
+    return self:GetNWInt("rvr_food")
+end
+
+function PlayerMeta:SetFood(n)
+    self:SetNWInt("rvr_food", n)
+end
+
+function PlayerMeta:AddFood(n)
```
Comment (samuelWilliams99) [inline gamemode/modules/hunger/sh_hunger.lua]: We'll also want a max hunger and thirst, probably 100

**Review by samuelWilliams99 — CHANGES_REQUESTED** [review]
:)

File: gamemode/config/hunger.lua [hunk 5]
```diff
@@ -0,0 +1,12 @@
+GM.Config = GM.Config or  {}
+
+local Config = GM.Config
+
+Config.MAX_FOOD = 100
+Config.MAX_WATER = 100
+Config.HUNGER_DELAY = 5
+Config.FOOD_LOSS_AMOUNT = -1
+Config.WATER_LOSS_AMOUNT = -1
+Config.HUNGER_DAMAGE_AMOUNT = 10
```
Comment (samuelWilliams99) [inline gamemode/config/hunger.lua]: Not a fan of having every config value in the gamemode in a flat table, would prefer a structure like `Config.Hunger.MAX_FOOD`, what do other people think?

Comment (overlordbleck) [reply gamemode/config/hunger.lua]: I like the thought of there being separate configs within the Config table

Comment (ghost) [reply gamemode/config/hunger.lua]: Is there no WATER_DAMAGE_AMOUNT ?

Comment (plally) [reply gamemode/config/hunger.lua]: hunger damage amount is shared for both of them right now

File: gamemode/modules/hud/cl_hud.lua [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    {
+        color = Color(181, 114, 60),
+        get = function()
+            return LocalPlayer():GetFood()
+        end,
+        max = 100,
+    },
+
+    {   name = "health",
```
Comment (plally) [inline gamemode/modules/hud/cl_hud.lua]:
```suggestion
    {  
```

File: gamemode/modules/hud/cl_hud.lua [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        end,
+        max = 100,
+    },
+
+    {
+        color = Color(181, 114, 60),
+        get = function()
+            return LocalPlayer():GetFood()
+        end,
+        max = 100,
```
Comment (plally) [inline gamemode/modules/hud/cl_hud.lua]:
```suggestion
        max = config.MAX_FOOD,
```

File: gamemode/modules/hud/cl_hud.lua [hunk 8]
```diff
@@ -0,0 +1,57 @@
+local stats = {
+    {
+        color = Color(138, 138, 230),
+        get  = function()
+            return  LocalPlayer():GetWater()
+        end,
+        max = 100,
```
Comment (plally) [inline gamemode/modules/hud/cl_hud.lua]:
```suggestion
        max = config.MAX_WATER,
```

File: gamemode/config/hunger.lua [hunk 9]
```diff
@@ -0,0 +1,13 @@
+GM.Config = GM.Config or  {}
+
+GM.Config.Hunger = {}
+local Config = GM.Config.Hunger
+
+Config.MAX_FOOD = 100
+Config.MAX_WATER = 100
+Config.DELAY = 0.5
+Config.LOSS_CHANCE = 0.1 -- 0 - 1
```
Comment (overlordbleck) [inline gamemode/config/hunger.lua:9]: 10% Chance to lose hunger?

File: gamemode/modules/hud/cl_hud.lua [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+    },
+}
+
+-- TODO widths and heights should scale with screen resolution
+function GM:HUDPaint()
+    local x = 0
+    local y = ScrH() - 5 - 35 * #stats
+    draw.RoundedBox( 2, x, y, 320, 35*#stats+5, Color(50, 50, 50, 190) )
```
Comment (overlordbleck) [inline gamemode/modules/hud/cl_hud.lua]:
```suggestion
    draw.RoundedBox( 2, x, y, 320, 35 * #stats + 5, Color(50, 50, 50, 190) ) 
```

**Review by overlordbleck — COMMENTED** [review]
There's a few places where you don't put whitespaces between parenthesis and arguments, but other than that it's looking good

File: gamemode/config/hunger.lua (hunk 9 above)
Comment (plally) [reply gamemode/config/hunger.lua:9]: yeah, we might want to lower that or raise the delay eventually

**Comment by overlordbleck** [comment]
Also, a good word for both hunger and thirst might be `appetite` or `nourishment`?

File: gamemode/modules/hud/cl_hud.lua [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    local x = 0
+    local y = ScrH() - 5 - 35 * #stats
+    draw.RoundedBox( 2, x, y, 320, 35 * #stats + 5, Color( 50, 50, 50, 190 ) ) 
+
+    for i, stat in ipairs( stats ) do
+        local width = 272
+        local height = 30
+        width = width * stat.get() / stat.max
+        local x = 40
+        local y = ScrH() - 35*i
```
Comment (overlordbleck) [inline gamemode/modules/hud/cl_hud.lua]:
```suggestion
        local y = ScrH() - 35 * i
```

File: gamemode/modules/hunger/sv_hunger.lua [hunk 12]
```diff
@@ -0,0 +1,36 @@
+local config = GM.Config.Hunger
+
+hook.Add( "PlayerSpawn", "RaftVRaft_HungerReset", function( ply, transition )
+    ply:SetFood( config.MAX_FOOD )
+    ply:SetWater( config.MAX_WATER )
+
```
Comment (overlordbleck) [inline gamemode/modules/hunger/sv_hunger.lua]: NEW LINE?

Comment (plally) [inline gamemode/modules/hunger/sv_hunger.lua]:
```suggestion
```

**Pull request merged**